File: text_style.go
package uipack

import (
	"bufio"
)

// A text style.
type TextStyle struct {
	FontFamily     string
	FontSize       float64
	FontWeight     uint8
	LetterSpacing  float64
	WordSpacing    float64
	LineHeight     float64
	FontVariations []FontVariation
}

func (v *TextStyle) Equal(o *TextStyle) bool {

	if v.FontSize != o.FontSize {
		return false
	}

	if v.FontWeight != o.FontWeight {
		return false
	}

	if v.LetterSpacing != o.LetterSpacing {
		return false
	}

	if v.WordSpacing != o.WordSpacing {
		return false
	}

	if v.LineHeight != o.LineHeight {
		return false
	}

	if len(v.FontVariations) != len(o.FontVariations) {
		return false
	}

	for i, variation := range v.FontVariations {
		ov := o.FontVariations[i]
		if variation.Axis != ov.Axis {
			return false
		}

		if variation.Value != ov.Value {
			return false
		}
	}
	return true
}

// A variation for variable fonts
type FontVariation struct {
	Axis  string
	Value float64
}

// Binary encoding

func (v *TextStyle) Encode(writer *bufio.Writer) error {
	err := writeString(writer, v.FontFamily)
	if err != nil {
		return err
	}
	err = writeFloat64(writer, v.FontSize)
	if err != nil {
		return err
	}
	err = writeUint8(writer, v.FontWeight)
	if err != nil {
		return err
	}
	err = writeFloat64(writer, v.LetterSpacing)
	if err != nil {
		return err
	}
	err = writeFloat64(writer, v.WordSpacing)
	if err != nil {
		return err
	}
	err = writeFloat64(writer, v.LineHeight)
	if err != nil {
		return err
	}

	err = writeUint32(writer, uint32(len(v.FontVariations)))
	if err != nil {
		return err
	}
	for _, variation := range v.FontVariations {
		err = variation.Encode(writer)
		if err != nil {
			return err
		}
	}
	return nil
}

func (v *TextStyle) Decode(reader *bufio.Reader) error {
	var err error
	v.FontFamily, err = readString(reader)
	if err != nil {
		return err
	}
	v.FontSize, err = readFloat64(reader)
	if err != nil {
		return err
	}
	v.FontWeight, err = readUint8(reader)
	if err != nil {
		return err
	}
	v.LetterSpacing, err = readFloat64(reader)
	if err != nil {
		return err
	}
	v.WordSpacing, err = readFloat64(reader)
	if err != nil {
		return err
	}
	v.LineHeight, err = readFloat64(reader)
	if err != nil {
		return err
	}

	variationsCount, err := readUint32(reader)
	if err != nil {
		return err
	}
	variations := make([]FontVariation, variationsCount)
	for i := uint32(0); i < variationsCount; i++ {
		variation := FontVariation{}
		err = variation.Decode(reader)
		variations[i] = variation
		if err != nil {
			return err
		}
	}
	v.FontVariations = variations
	return nil
}

func (v *FontVariation) Encode(writer *bufio.Writer) error {
	err := writeString(writer, v.Axis)
	if err != nil {
		return err
	}
	err = writeFloat64(writer, v.Value)
	if err != nil {
		return err
	}
	return nil
}

func (v *FontVariation) Decode(reader *bufio.Reader) error {
	var err error
	v.Axis, err = readString(reader)
	if err != nil {
		return err
	}
	v.Value, err = readFloat64(reader)
	if err != nil {
		return err
	}

	return nil
}

File: packages/uipack/lib/src/models/text_style.dart
import '../encoding/binary_encoding.dart';

class FontVariation {
  String axis;
  double value;

  FontVariation({required this.axis, required this.value});

  void encode(ByteDataWriter writer) {
    writer.writeString(axis);
    writer.writeFloat64(value);
  }

  void decode(ByteDataReader reader) {
    axis = reader.readString();
    value = reader.readFloat64();
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    return other is FontVariation && 
           other.axis == axis && 
           other.value == value;
  }

  @override
  int get hashCode => axis.hashCode ^ value.hashCode;
}

class TextStyle {
  String fontFamily;
  double fontSize;
  int fontWeight;
  double letterSpacing;
  double wordSpacing;
  double lineHeight;
  List<FontVariation> fontVariations;

  TextStyle({
    required this.fontFamily,
    required this.fontSize,
    required this.fontWeight,
    required this.letterSpacing,
    required this.wordSpacing,
    required this.lineHeight,
    required this.fontVariations,
  });

  void encode(ByteDataWriter writer) {
    writer.writeString(fontFamily);
    writer.writeFloat64(fontSize);
    writer.writeUint8(fontWeight);
    writer.writeFloat64(letterSpacing);
    writer.writeFloat64(wordSpacing);
    writer.writeFloat64(lineHeight);
    
    writer.writeUint32(fontVariations.length);
    for (final variation in fontVariations) {
      variation.encode(writer);
    }
  }

  void decode(ByteDataReader reader) {
    fontFamily = reader.readString();
    fontSize = reader.readFloat64();
    fontWeight = reader.readUint8();
    letterSpacing = reader.readFloat64();
    wordSpacing = reader.readFloat64();
    lineHeight = reader.readFloat64();
    
    final variationsCount = reader.readUint32();
    fontVariations = List.generate(variationsCount, (index) {
      final variation = FontVariation(axis: '', value: 0.0);
      variation.decode(reader);
      return variation;
    });
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    return other is TextStyle &&
           other.fontFamily == fontFamily &&
           other.fontSize == fontSize &&
           other.fontWeight == fontWeight &&
           other.letterSpacing == letterSpacing &&
           other.wordSpacing == wordSpacing &&
           other.lineHeight == lineHeight &&
           _listEquals(other.fontVariations, fontVariations);
  }

  bool _listEquals<T>(List<T> a, List<T> b) {
    if (a.length != b.length) return false;
    for (int i = 0; i < a.length; i++) {
      if (a[i] != b[i]) return false;
    }
    return true;
  }

  @override
  int get hashCode => Object.hash(
    fontFamily,
    fontSize,
    fontWeight,
    letterSpacing,
    wordSpacing,
    lineHeight,
    Object.hashAll(fontVariations),
  );

  @override
  String toString() => 'TextStyle(family: $fontFamily, size: $fontSize, weight: $fontWeight)';
}